Webui/server config defaults

ServeurpersoCom opened the pull request:
*Make sure to read the [contributing guidelines](https://github.com/ggml-org/llama.cpp/blob/master/CONTRIBUTING.md) before submitting a PR*

## server/webui: add CLI support for configuring WebUI defaults

### Summary

Implements issue #17940: adds '--webui-config-file' CLI argument and 'LLAMA_WEBUI_CONFIG' environment variable to configure WebUI default settings from the server side.

### Problem

WebUI settings are hardcoded in the frontend with no server-side configuration. This causes problems:

- Server operators can't set sensible defaults for their deployment
- Users must reconfigure settings every time localStorage is cleared
- Multi-user servers can't enforce defaults (like disabling auto-file-paste)

Example: Issue #17940 requests ability to set 'pasteLongTextToFileLen: 0' by default, but currently every user must manually configure this in every new browser session.

### The TODO that guided this implementation

While exploring the codebase, I found this intentional placeholder in 'server-models.cpp' (lines 6-16):

```cpp
// TODO: add support for this on web UI
{"role", "router"},
// this is a dummy response to make sure webui doesn't break
{"default_generation_settings", {
    {"params", json{}},  // <-- Empty JSON waiting for implementation
    {"n_ctx", 0},
}},
```

This empty JSON was left by the original developers as a clear signal for where webui settings should go. This PR completes that TODO.

## Backend Changes (C++)

### What was added

**1. CLI argument** ('common/arg.cpp')
- New flag: '--webui-config-file <path>'
- Loads JSON file with WebUI default settings

**2. Environment variable** ('server.cpp')
- New env var: 'LLAMA_WEBUI_CONFIG'
- Accepts JSON string, overrides file config
- Example: 'LLAMA_WEBUI_CONFIG='{"pasteLongTextToFileLen":0}''

**3. JSON parsing logic** ('server.cpp')
- Parses config at startup, before any router/child fork
- Simple override: env var OR file (env var takes precedence)
- Error handling: warn on missing file, error on invalid JSON

**4. Expose in /props** ('server-models.cpp' + 'server-context.cpp')
- Added 'webui_settings' field to '/props' response
- Works in both router mode and child mode
- Fills in the TODO placeholder mentioned above

### How it works

```
Server startup
  |
  v
Parse LLAMA_WEBUI_CONFIG env var (if set) OR config.json (if provided)
  |
  v
Store result in params.webui_config
  |
  v
Return in /props as webui_settings
```

The C++ backend is completely agnostic: it doesn't know what settings exist. It just:
- Reads JSON
- Passes through to frontend

This means adding new WebUI settings never requires C++ changes.

## Frontend Changes (TypeScript)

### What was added

**1. API type** ('types/api.d.ts')
- Added optional 'webui_settings' field to props response
- Type: 'Record<string, string | number | boolean>'

**2. Syncable parameters** ('services/parameter-sync.ts')
- Added 14 WebUI settings to 'SYNCABLE_PARAMETERS':
  - pasteLongTextToFileLen
  - pdfAsImage
  - showThoughtInProgress
  - showToolCalls
  - disableReasoningFormat
  - keepStatsVisible
  - showMessageStats
  - askForTitleConfirmation
  - disableAutoScroll
  - renderUserContentAsMarkdown
  - autoMicOnEmpty
  - pyInterpreterEnabled
  - enableContinueGeneration
  - showSystemMessage
- Updated 'extractServerDefaults()' to merge webui_settings

**3. Store integration** ('stores/server.svelte.ts' + 'stores/settings.svelte.ts')
- Added getter for 'webuiSettings'
- Plumbed through to settings store

**4. Tests** ('parameter-sync.spec.ts')
- Test for webui_settings extraction
- Validates user-specific settings are filtered out
- Uses 'pasteLongTextToFileLen: 0' example from issue

### Settings intentionally excluded

These are user-specific and should NOT be server-configurable:
- apiKey (personal credentials)
- systemMessage (personal prompt)
- theme (UI preference)
- mcpServers (personal MCP config)
- custom (user's custom JSON)

## Priority Order

Server operators set defaults, but users keep control:

```
1. WebUI hardcoded defaults (settings-config.ts)
2. Server config: LLAMA_WEBUI_CONFIG env var OR --webui-config-file
3. User's localStorage (always wins)
```

This means:
- Server can provide sensible defaults
- Users can still override anything
- No breaking changes

## Usage Examples

### Config file
```bash
# webui-defaults.json
{
  "pasteLongTextToFileLen": 0,
  "pdfAsImage": true,
  "showThoughtInProgress": false
}

./llama-server --webui-config-file webui-defaults.json
```

### Environment variable
```bash
LLAMA_WEBUI_CONFIG='{"pasteLongTextToFileLen":0}' ./llama-server
```

### Both (env var overrides file)
```bash
LLAMA_WEBUI_CONFIG='{"pdfAsImage":true}' \
  ./llama-server --webui-config-file webui-defaults.json
```

### Check it works
```bash
curl http://localhost:8080/props | jq .webui_settings
```

## Testing

### Backend
```bash
# Create test config
echo '{"pasteLongTextToFileLen":0}' > test.json

# Test file loading
./llama-server --webui-config-file test.json

# Test env var
LLAMA_WEBUI_CONFIG='{"pdfAsImage":true}' ./llama-server

# Check /props
curl http://localhost:8080/props | jq .webui_settings
```

### Frontend
```bash
cd tools/server/webui
npm test -- parameter-sync.spec.ts
```

## Backward Compatibility

No breaking changes:
- 'webui_settings' is optional in /props
- Old servers work fine (webui_settings: null or missing)
- WebUI falls back to hardcoded defaults
- No recompilation needed for existing deployments

## What this enables

Now server operators can:
- Set deployment-wide defaults without forking
- Disable problematic features (like auto-file-paste)
- Provide better defaults for their use case
- Support multiple configurations via env vars

And adding new configurable settings only requires TypeScript changes: no C++ recompilation needed.

Fixes #17940

ServeurpersoCom commented on common/common.h:
```diff
@@ -12,6 +12,8 @@
 #include <vector>
 #include <map>
 
+#include <nlohmann/json.hpp>
```
Done: changed to std::string webui_config_json in common.h and parse locally in server-context/server-models where json.hpp is already included. Thanks

ggerganov commented on common/common.h:
```diff
@@ -482,6 +482,8 @@ struct common_params {
+    std::string webui_config_json;
     bool endpoint_slots   = true;
     bool endpoint_props   = false; // only control POST requests, not GET
     bool endpoint_metrics = false;
```
You can arrange these a bit better like so:

```suggestion
    // webui configs
    bool webui = true;
    
    std::string webui_config_file;
    std::string webui_config_json;
    
    // "advanced" endpoints are disabled by default for better security
    bool endpoint_slots   = true;
    bool endpoint_props   = false; // only control POST requests, not GET
    bool endpoint_metrics = false;
```

ggerganov commented on common/arg.cpp:
```diff
@@ -2493,6 +2493,13 @@ common_params_context common_params_parser_init(common_params & params, llama_ex
+        [](common_params & params, const std::string & value) {
+            params.webui_config_file = value;
+        }
+    ).set_examples({LLAMA_EXAMPLE_SERVER}));
```
I think you need to use `.set_env("LLAMA_ARG_WEBUI_CONFIG_FILE")` here

ServeurpersoCom commented on tools/server/server.cpp:
```diff
@@ -2,19 +2,24 @@
 #include "server-http.h"
 #include "server-models.h"
 
+#include "server-common.h"
```
Cleaner !

allozaur reviewed (approved):
Looking good on my end

ggerganov reviewed (approved):
Can merge after rebase and green CI. @ngxson up to you

ngxson commented on tools/server/server-models.cpp:
```diff
@@ -811,6 +811,7 @@ void server_models_routes::init_routes() {
         if (name.empty()) {
             // main instance
             auto res = std::make_unique<server_http_res>();
+            json webui_settings = params.webui_config_json.empty() ? json::object() : json::parse(params.webui_config_json);
```
calling `json::parse` here will potentially impact the performance if this endpoint is repeatedly called by the client (which is the case in practice)

I would recommend to parse this is `server_context::webui_config` inside `server_context::init()` and reuse it later

ngxson commented:
Also, you need to re-generate the server docs via `llama-gen-docs`, and copy-paste it to the `## Usage` section of server/README.md

ServeurpersoCom commented on tools/server/server.cpp:
```diff
@@ -8,13 +9,17 @@
 #include <windows.h>
 #endif
 
+using json = nlohmann::ordered_json;
```
Removed: json handling is now only in `server_context` and `server_models_routes`.

ngxson commented on tools/server/server.cpp:
```diff
@@ -1,3 +1,4 @@
+#include "server-common.h"
```
this is probably redundant

```suggestion
```

ngxson reviewed (approved):
Can be merged once you update the docs 👍

ServeurpersoCom committed:
server/webui: add server-side WebUI config support

Add CLI arguments --webui-config (inline JSON) and --webui-config-file
(file path) to configure WebUI default settings from server side.

Backend changes:
- Parse JSON once in server_context::load_model() for performance
- Cache parsed config in webui_settings member (zero overhead on /props)
- Add proper error handling in router mode with try/catch
- Expose webui_settings in /props endpoint for both router and child modes

Frontend changes:
- Add 14 configurable WebUI settings via parameter sync
- Add tests for webui settings extraction
- Fix subpath support with base path in API calls

Addresses feedback from @ngxson and @ggerganov

ServeurpersoCom committed:
server: address review feedback from ngxson

ServeurpersoCom committed:
server: regenerate README with llama-gen-docs

jbone1313 commented:
@ServeurpersoCom this is not working for me. I verified that the props page has the correct settings and ensured that they actually change when set via the config file. But, the settings in the web UI do not match the props, nor does the web UI behave according to what is set in the props page.

b7472
Windows

ServeurpersoCom commented:
> @ServeurpersoCom this is not working for me. I verified that the props page has the correct settings and ensured that they actually change when set via the config file. But, the settings in the web UI do not match the props, nor does the web UI behave according to what is set in the props page.
> 
> b7472 Windows

Your previous WebUI settings can override the server defaults. Server defaults only apply when you first visit the WebUI or after resetting settings.
<img width="813" height="519" alt="Sans titre" src="https://github.com/user-attachments/assets/b1a3e93b-0a45-4d0d-9053-0c6f99ef9e68" />

Else can you clear your browser's localStorage and retry ?
<img width="193" height="52" alt="2" src="https://github.com/user-attachments/assets/54be0223-3df6-4161-ad81-7557336693ed" />

jbone1313 commented:
Yeah, I just tried "reset to default" and the code you gave me for Firefox. Neither worked. After I cleared all browser data it worked. 

While I am here:

1. Is there a way to force the server defaults to override the previous Web UI settings without having to reset? It would be nice to know that what is set in the defaults will win every time. 
2. Is there a list of the actual names of the settable parameters? I already checked -h and the llama-server readme. The list you posted up-thread has many but not all of them. For example, the new "always show sidebar parameter" is not in that list.

jbone1313 commented:
RE 1:

I am happy to write up a feature request, but after thinking about this, it seems like this feature needs to be implemented consistently with how this issue will be handled: https://github.com/ggml-org/llama.cpp/issues/18129#issue-3738125194 (Misc. bug: Web-Frontend overrides sampling parameters from models preset in router mode), and l imagine that one will be treated with high priority. 

I am not a UI expert, but it seems like some elegant and consistent solution could apply to both cases. 

Perhaps it could work thus: for a given parameter, the server (either server config defaults or router config.ini) - IF set - is the source of truth unless the parameter is manually set in Web UI. Perhaps the Web UI could show the server-set values as "greyed-out," and when they are manually overridden, they would be shown as "normal." This requires no extra JSON fields, and the UI itself would provide information about where the parameter values are being set. 

RE "prohibiting overrides" for enterprise-like cases: perhaps that could be considered low priority and implemented later.

RE 2:

Apologies, but I do not see how that addresses my question 2 above regarding how to determine the actual list of settable parameters we can set in the JSON file.

ServeurpersoCom commented:
@jbone1313 What you're describing is exactly the design pattern we need for #18129 (sampling params override bug). Server values as source of truth (greyed in UI), user overrides saved per-model. Same architecture for both webui settings and sampling parameters. We all independently converged on this.

I've been running shared llama.cpp servers for friends for a long time and wanted this exact behavior. Now deploying more and more at work: proper server governance is critical.
